File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/InputStreamFlowableSubscriber.java
package net.ballmerlabs.uscatterbrain.network;

import org.reactivestreams.Subscription;

import java.io.IOException;

import io.reactivex.FlowableSubscriber;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;

public class InputStreamFlowableSubscriber extends InputStreamCallback
        implements FlowableSubscriber<byte[]> {
    private boolean isDisposed = false;
    private Subscription subscription;
    private int blocksize;
    public static final int DEFAULT_BLOCKSIZE = 20;

    public InputStreamFlowableSubscriber() {
        this.blocksize = DEFAULT_BLOCKSIZE;
    }

    public InputStreamFlowableSubscriber(int blocksize) {
        this.blocksize = blocksize;
    }

    public void setBlocksize(int blocksize) {
        this.blocksize = blocksize;
    }

    @Override
    public void onSubscribe(@NonNull Subscription s) {
        s.request(blocksize*20);
        subscription = s;
        disposable = new Disposable() {
            @Override
            public void dispose() {
                isDisposed = true;
                s.cancel();
            }

            @Override
            public boolean isDisposed() {
                return isDisposed;
            }
        };
    }

    @Override
    public void onNext(byte[] bytes) {
        if (!closed) {
            acceptBytes(bytes);
        }
    }

    @Override
    public void onError(Throwable e) {
        throwable = e;
        try {
            close();
        } catch (IOException ignored) {

        }
    }

    @Override
    public void onComplete() {

    }

    @Override
    public int read(byte[] b) throws IOException {
        if (subscription != null) {
            subscription.request(Math.max(blocksize, b.length));
        }
        return super.read(b);
    }

    @Override
    public int read(byte[] b, int off, int len) throws IOException {
        if (subscription != null) {
            subscription.request(Math.max(blocksize, Math.min(b.length, len)));
        }
        return super.read(b, off, len);
    }

    @Override
    public int read() throws IOException {
        if (subscription != null) {
            subscription.request(1);
        }
        return super.read();
    }


    @Override
    public long skip(long n) throws IOException {
        if (subscription != null) {
            subscription.request(Math.max(blocksize, n));
        }
        return super.skip(n);
    }

    @Override
    public void close() throws IOException {
        super.close();
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/InputStreamFlowableSubscriber.kt
package net.ballmerlabs.uscatterbrain.network

import io.reactivex.FlowableSubscriber
import io.reactivex.disposables.Disposable
import org.reactivestreams.Subscription
import java.io.IOException
import kotlin.math.max
import kotlin.math.min

/**
 * serves as a bridge between a flowable and an inputstream with
 * a static sized buffer to old unread data
 */
class InputStreamFlowableSubscriber(capacity: Int) : InputStreamCallback(capacity), FlowableSubscriber<ByteArray?> {
    private var isDisposed = false
    private var subscription: Subscription? = null
    private var blocksize: Int = DEFAULT_BLOCKSIZE

    override fun onSubscribe(s: Subscription) {
        s.request(blocksize * 20.toLong())
        subscription = s
    }

    override fun onNext(t: ByteArray?) {
        if (!closed) {
            acceptBytes(t!!)
        }
    }

    override fun onError(e: Throwable) {
        throwable = e
        try {
            close()
        } catch (ignored: IOException) {
        }
    }

    override fun onComplete() {}

    @Throws(IOException::class)
    override fun read(b: ByteArray): Int {
        subscription?.request(max(blocksize, b.size).toLong())
        return super.read(b)
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray, off: Int, len: Int): Int {
        subscription?.request(max(blocksize, min(b.size, len)).toLong())
        return super.read(b, off, len)
    }

    @Throws(IOException::class)
    override fun read(): Int {
        subscription?.request(1)
        return super.read()
    }

    @Throws(IOException::class)
    override fun skip(n: Long): Long {
        subscription?.request(max(blocksize.toLong(), n))
        return super.skip(n)
    }

    @Throws(IOException::class)
    override fun close() {
        super.close()
    }

    companion object {
        const val DEFAULT_BLOCKSIZE = 20
    }
}